### PR TITLE
feat: Use nest-asyncio when configured with env var

### PR DESCRIPTION
**Pull request by KShivendu**
## Summary of change

Use nest-asyncio when configured with env var

## Checklist for important updates

-   [x] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens_python/constants.py`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [x] Changes to the version if needed
    -   In `setup.py`
    -   In `supertokens_python/constants.py`
-   [x] Had installed and ran the pre-commit hook
-   [x] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `supertokens_python/utils.py` file to include that in the `FRAMEWORKS` variable
-   [ ] If added a new recipe that has a User type with extra info, then be sure to change the User type in supertokens_python/types.py
-   [ ] Make sure that `syncio` / `asyncio` functions are consistent.
-   [ ] If access token structure has changed
    -   Modified test in `tests/sessions/test_access_token_version.py` to account for any new claims that are optional or omitted by the core

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [unreleased]
 
+## [0.16.2] - 2023-09-20
+
+- Allow use of [nest-asyncio](https://pypi.org/project/nest-asyncio/) when env var `SUPERTOKENS_NEST_ASYNCIO=1`.
```
Comment (KShivendu) [inline CHANGELOG.md:13]: Env var is better than changing the SDK interface. This is anyways a rare requirement and this way we can keep python SDK consistent with others.

File: supertokens_python/utils.py [hunk 2]
```diff
@@ -195,28 +192,6 @@ def find_first_occurrence_in_list(
...
-            "Inconsistent mode detected, check if you are using the right asgi / wsgi mode",
-            category=RuntimeWarning,
-        )
-
-    if real_mode == "wsgi":
-        asyncio.run(func())
-    else:
-        check_event_loop()
-        loop = asyncio.get_event_loop()
-        loop.create_task(func())
```
Comment (KShivendu) [inline supertokens_python/utils.py:217]: This function wasn't used anywhere throughout the code. We discussed removing this last time.

File: supertokens_python/async_to_sync_wrapper.py [hunk 3]
```diff
@@ -14,20 +14,37 @@
...
     try:
-        asyncio.get_event_loop()
-    except RuntimeError as ex:
+        import nest_asyncio as _  # type: ignore
+
+        return getenv("SUPERTOKENS_NEST_ASYNCIO", "") == "1"
+    except Exception:
+        return False
+
+
```
Comment (rishabhpoddar) [inline supertokens_python/async_to_sync_wrapper.py:30]: make sure that there is at least one thing in cicd where we run the python sdk (for example in e2e test with flask), where nest_asyncio is not a dependency and then we make sure it all works

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: tests/test_config.py [hunk 4]
```diff
@@ -800,3 +802,26 @@ async def test_cookie_samesite_with_ec2_public_url():
...
+        # nest-asyncio should be installed
+        sync(asyncio.sleep(0.1))
+    else:
+        # nest-asyncio shouldn't be installed and sync() should throw error
+        try:
+            sync(asyncio.sleep(0.1))
+            assert False, "Shouldn't come here"
+        except ModuleNotFoundError:
+            # should be missing
+            assert True
```
Comment (rishabhpoddar) [inline tests/test_config.py:827]: remove this

Comment (KShivendu) [reply tests/test_config.py]: Done

**Pull request merged**